front/loadObjects.js

```diff
@@ -1,44 +1,118 @@
+var unpackBuilding = require('./unpackBuilding.js');
+var infosFromMesh = require('./infosFromMesh.js');
+var meshFromId = require('./meshFromId.js');
+var LimitedEntryMap = require('./LimitedEntryMap.js');
+// var meshColor = require('../common/meshDefaultColor.js');
 
+function distance(a, b){
+    return Math.sqrt((a.x - b.x)*(a.x - b.x) + (a.y - b.y)*(a.y - b.y));
+}
 
-module.exports = function(fetchCityObject){
-    
-    return function loadObjects(scene, south, north, east, west) {
+module.exports = function(getCityObject){
+
+    function Fast(scene, point, distance) {
         // query the rtree to know what building are needed
+
+        var nbToLoad = 0;
+
+        var south = point.y - distance;
+        var north = point.y + distance;
+        var west = point.x - distance;
+        var east = point.x + distance;
+
         var results = rTree.search([west, south, east, north]);
 
-        var objectToDisplayIds = new Set(results.map(function(r){ return r[4].id; }));
-        var displayedObjectIds = new Set(buildingMap.keys().filter(function(id){ return buildingMap.get(id).visible; }));
+        var objectRequestedIds = new Set(results.map(function(r){ return r[4].id; }));
+
+        // in requested zone ...
+        objectRequestedIds.forEach(function(id){
 
-        // add objects which should be visible but aren't
-        objectToDisplayIds.forEach(function(id){
-            if(!buildingMap.has(id)){
-                fetchCityObject(id); // object will be added to scene in serverCommunication.js, which is a terrible place, yes.
+            // ... and in meshFromId (is already in scene) => do nothing
+
+            // ... and not in meshFromId (not in scene) => request building
+            if (!meshFromId.has(id)){
+                nbToLoad ++;
+
+                getCityObject(id)
+                .then(function(object){
+                    var mesh = unpackBuilding(object);
+                    scene.add(mesh);
+                });
             }
-            else{
-                var entry = buildingMap.get(id);
+                
+        });
 
-                if(!entry.visible){
-                    scene.add(entry.mesh);
-                    entry.visible = true;
-                }
+        console.log('nb Visible', meshFromId.size, 'requestedIds', objectRequestedIds.size, 'nbToLoad', nbToLoad);
+    };
+    
+    function Zone(scene, camera, domElement) {
+        // query the rtree to know what building are needed
+        var nbToLoad = 0;
+
+        var position = camera.position;
+
+        var L = 2 * camera.position.z * Math.tan(Math.PI*camera.fov/(2*180));
+        var l = L * domElement.clientWidth / domElement.clientHeight;
+
+        var south = position.y - L/2;
+        var north = position.y + L/2;
+        var west = position.x - l/2;
+        var east = position.x + l/2;
+        
+        // ask for a little extra
+        west -= 100;
+        south -= 100;
+        east += 100;
+        north += 100;
```

Comment:
It's not the role of the API to hardcode these values. This should happen in client code.